### PR TITLE
[DRAFT] guacamole: init at 1.4.0

### PR DESCRIPTION
**Pull request by onny**
###### Description of changes

Continuation of https://github.com/NixOS/nixpkgs/pull/48140 without the module, only the package. Addresses https://github.com/NixOS/nixpkgs/issues/17879

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by InLaw** [comment]
- system: `"x86_64-linux"` 
 - host os: `Linux 5.15.64, NixOS, 22.05 (Quokka), 22.05.2992.45b56b5321a` 
 - multi-user?: `yes` 
 - sandbox: `yes` 
 - version: `nix-env (Nix) 2.8.1` 
 - channels(root): `"home-manager-22.05.tar.gz, nixos-22.05"` 
 - channels(usera): `""` 
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`


currently it results in 

```
------------------------------------------------
guacamole-server version 1.4.0
------------------------------------------------

   Library status:

     freerdp2 ............ yes
     pango ............... yes
     libavcodec .......... no
     libavformat.......... no
     libavutil ........... no
     libssh2 ............. yes
     libssl .............. yes
     libswscale .......... no
     libtelnet ........... yes
     libVNCServer ........ yes
     libvorbis ........... yes
     libpulse ............ yes
     libwebsockets ....... no
     libwebp ............. yes
     wsock32 ............. no

   Protocol support:

      Kubernetes .... no
      RDP ........... yes
      SSH ........... yes
      Telnet ........ yes
      VNC ........... yes

   Services / tools:

      guacd ...... yes
      guacenc .... no
      guaclog .... yes

   FreeRDP plugins: /nix/store/mh54ww1apv755w84x2cgdclhf2c42dsh-freerdp-2.8.0/lib/freerdp2
   Init scripts: no
   Systemd units: no

Type "make" to compile guacamole-server.

@nix { "action": "setPhase", "phase": "buildPhase" }
building
build flags: SHELL=/nix/store/p7bpdnxqd3i5hwm92mrscf7mvxk66404-bash-5.1-p16/bin/bash
make  all-recursive
make[1]: Entering directory '/build/source'
Making all in src/libguac
make[2]: Entering directory '/build/source/src/libguac'
Making all in .
make[3]: Entering directory '/build/source/src/libguac'
  CC       libguac_la-argv.lo
  CC       libguac_la-audio.lo
  CC       libguac_la-client.lo
  CC       libguac_la-encode-jpeg.lo
  CC       libguac_la-encode-png.lo
  CC       libguac_la-error.lo
  CC       libguac_la-hash.lo
  CC       libguac_la-id.lo
  CC       libguac_la-palette.lo
  CC       libguac_la-parser.lo
  CC       libguac_la-pool.lo
  CC       libguac_la-protocol.lo
  CC       libguac_la-raw_encoder.lo
  CC       libguac_la-socket.lo
  CC       libguac_la-socket-broadcast.lo
  CC       libguac_la-socket-fd.lo
  CC       libguac_la-socket-nest.lo
  CC       libguac_la-socket-tee.lo
  CC       libguac_la-string.lo
  CC       libguac_la-timestamp.lo
  CC       libguac_la-unicode.lo
  CC       libguac_la-user.lo
  CC       libguac_la-user-handlers.lo
  CC       libguac_la-user-handshake.lo
  CC       libguac_la-wait-fd.lo
  CC       libguac_la-wol.lo
  CC       libguac_la-encode-webp.lo
  CC       libguac_la-socket-ssl.lo
  CCLD     libguac.la
make[3]: Leaving directory '/build/source/src/libguac'
Making all in tests
make[3]: Entering directory '/build/source/src/libguac/tests'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/build/source/src/libguac/tests'
make[2]: Leaving directory '/build/source/src/libguac'
Making all in src/common
make[2]: Entering directory '/build/source/src/common'
Making all in .
make[3]: Entering directory '/build/source/src/common'
  CC       libguac_common_la-io.lo
  CC       libguac_common_la-blank_cursor.lo
  CC       libguac_common_la-clipboard.lo
  CC       libguac_common_la-cursor.lo
  CC       libguac_common_la-display.lo
  CC       libguac_common_la-dot_cursor.lo
  CC       libguac_common_la-ibar_cursor.lo
  CC       libguac_common_la-iconv.lo
  CC       libguac_common_la-json.lo
  CC       libguac_common_la-list.lo
  CC       libguac_common_la-pointer_cursor.lo
  CC       libguac_common_la-recording.lo
  CC       libguac_common_la-rect.lo
  CC       libguac_common_la-string.lo
  CC       libguac_common_la-surface.lo
  CCLD     libguac_common.la
make[3]: Leaving directory '/build/source/src/common'
Making all in tests
make[3]: Entering directory '/build/source/src/common/tests'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/build/source/src/common/tests'
make[2]: Leaving directory '/build/source/src/common'
Making all in src/common-ssh
make[2]: Entering directory '/build/source/src/common-ssh'
Making all in .
make[3]: Entering directory '/build/source/src/common-ssh'
  CC       libguac_common_ssh_la-buffer.lo
  CC       libguac_common_ssh_la-dsa-compat.lo
  CC       libguac_common_ssh_la-rsa-compat.lo
  CC       libguac_common_ssh_la-sftp.lo
  CC       libguac_common_ssh_la-ssh.lo
  CC       libguac_common_ssh_la-key.lo
key.c: In function 'guac_common_ssh_key_alloc':
key.c:63:9: error: 'PEM_read_bio_RSAPrivateKey' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
   63 |         rsa_key = PEM_read_bio_RSAPrivateKey(key_bio, NULL, NULL, passphrase);
      |         ^~~~~~~
In file included from key.c:33:
/nix/store/gd8xbg6y246ri68pz4qkwdczydwm6liq-openssl-3.0.5-dev/include/openssl/pem.h:447:1: note: declared here
  447 | DECLARE_PEM_rw_cb_attr(OSSL_DEPRECATEDIN_3_0, RSAPrivateKey, RSA)
      | ^~~~~~~~~~~~~~~~~~~~~~
key.c:79:9: error: 'RSA_get0_key' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
   79 |         RSA_get0_key(rsa_key, &key_n, &key_e, NULL);
      |         ^~~~~~~~~~~~
In file included from common-ssh/rsa-compat.h:26,
                 from key.c:25:
/nix/store/gd8xbg6y246ri68pz4qkwdczydwm6liq-openssl-3.0.5-dev/include/openssl/rsa.h:217:28: note: declared here
  217 | OSSL_DEPRECATEDIN_3_0 void RSA_get0_key(const RSA *r,
      |                            ^~~~~~~~~~~~
key.c:105:9: error: 'PEM_read_bio_DSAPrivateKey' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  105 |         dsa_key = PEM_read_bio_DSAPrivateKey(key_bio, NULL, NULL, passphrase);
      |         ^~~~~~~
In file included from key.c:33:
/nix/store/gd8xbg6y246ri68pz4qkwdczydwm6liq-openssl-3.0.5-dev/include/openssl/pem.h:453:1: note: declared here
  453 | DECLARE_PEM_rw_cb_attr(OSSL_DEPRECATEDIN_3_0, DSAPrivateKey, DSA)
      | ^~~~~~~~~~~~~~~~~~~~~~
key.c:121:9: error: 'DSA_get0_pqg' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  121 |         DSA_get0_pqg(dsa_key, &key_p, &key_q, &key_g);
      |         ^~~~~~~~~~~~
In file included from common-ssh/dsa-compat.h:26,
                 from key.c:23:
/nix/store/gd8xbg6y246ri68pz4qkwdczydwm6liq-openssl-3.0.5-dev/include/openssl/dsa.h:201:28: note: declared here
  201 | OSSL_DEPRECATEDIN_3_0 void DSA_get0_pqg(const DSA *d, const BIGNUM **p,
      |                            ^~~~~~~~~~~~
key.c:122:9: error: 'DSA_get0_key' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  122 |         DSA_get0_key(dsa_key, &pub_key, NULL);
      |         ^~~~~~~~~~~~
In file included from common-ssh/dsa-compat.h:26,
                 from key.c:23:
/nix/store/gd8xbg6y246ri68pz4qkwdczydwm6liq-openssl-3.0.5-dev/include/openssl/dsa.h:204:28: note: declared here
  204 | OSSL_DEPRECATEDIN_3_0 void DSA_get0_key(const DSA *d, const BIGNUM **pub_key,
      |                            ^~~~~~~~~~~~
key.c: In function 'guac_common_ssh_key_free':
key.c:164:9: error: 'RSA_free' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  164 |         RSA_free(key->rsa);
      |         ^~~~~~~~
In file included from common-ssh/rsa-compat.h:26,
                 from key.c:25:
/nix/store/gd8xbg6y246ri68pz4qkwdczydwm6liq-openssl-3.0.5-dev/include/openssl/rsa.h:293:28: note: declared here
  293 | OSSL_DEPRECATEDIN_3_0 void RSA_free(RSA *r);
      |                            ^~~~~~~~
key.c:166:9: error: 'DSA_free' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  166 |         DSA_free(key->dsa);
      |         ^~~~~~~~
In file included from common-ssh/dsa-compat.h:26,
                 from key.c:23:
/nix/store/gd8xbg6y246ri68pz4qkwdczydwm6liq-openssl-3.0.5-dev/include/openssl/dsa.h:127:28: note: declared here
  127 | OSSL_DEPRECATEDIN_3_0 void DSA_free(DSA *r);
      |                            ^~~~~~~~
key.c: In function 'guac_common_ssh_key_sign':
key.c:202:13: error: 'RSA_sign' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  202 |             if (RSA_sign(NID_sha1, digest, dlen, sig, &len, key->rsa) == 1)
      |             ^~
In file included from common-ssh/rsa-compat.h:26,
                 from key.c:25:
/nix/store/gd8xbg6y246ri68pz4qkwdczydwm6liq-openssl-3.0.5-dev/include/openssl/rsa.h:348:27: note: declared here
  348 | OSSL_DEPRECATEDIN_3_0 int RSA_sign(int type, const unsigned char *m,
      |                           ^~~~~~~~
key.c:208:13: error: 'DSA_do_sign' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  208 |             DSA_SIG* dsa_sig = DSA_do_sign(digest, dlen, key->dsa);
      |             ^~~~~~~
In file included from common-ssh/dsa-compat.h:26,
                 from key.c:23:
/nix/store/gd8xbg6y246ri68pz4qkwdczydwm6liq-openssl-3.0.5-dev/include/openssl/dsa.h:113:32: note: declared here
  113 | OSSL_DEPRECATEDIN_3_0 DSA_SIG *DSA_do_sign(const unsigned char *dgst, int dlen,
      |                                ^~~~~~~~~~~
cc1: all warnings being treated as errors
make[3]: *** [Makefile:596: libguac_common_ssh_la-key.lo] Error 1
make[3]: Leaving directory '/build/source/src/common-ssh'
make[2]: *** [Makefile:618: all-recursive] Error 1
make[2]: Leaving directory '/build/source/src/common-ssh'
make[1]: *** [Makefile:539: all-recursive] Error 1
make[1]: Leaving directory '/build/source'
make: *** [Makefile:461: all] Error 2

```

File: pkgs/servers/guacamole/default.nix [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+  propogatedBuildInputs = [
+    freerdp
+    autoreconfHook
+    pkg-config
+    cairo
+    libpng
+    libjpeg_turbo
+    libossp_uuid
+    freerdp
```
Comment (dasj19) [inline pkgs/servers/guacamole/default.nix:67]: this appears twice. this line should be removed

File: pkgs/servers/guacamole/default.nix [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    libvncserver
+    libpulseaudio
+    openssl
+    libvorbis
+    libwebp
+    inetutils
+  ];
+
+  patchPhase = ''
+    substituteInPlace ./src/protocols/rdp/keymaps/generate.pl --replace /usr/bin/perl "${perl}/bin/perl"
```
Comment (dasj19) [inline pkgs/servers/guacamole/default.nix:79]: this substitute is not needed anymore (version 1.4.0)

File: pkgs/servers/guacamole/default.nix [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    substituteInPlace ./src/protocols/rdp/Makefile.am --replace "-Werror -Wall" "-Wall"
+  '';
```
Comment (dasj19) [inline pkgs/servers/guacamole/default.nix:81]: I also needed to add the following:
```
substituteInPlace ./configure.ac --replace "FREERDP2_PLUGIN_DIR=" "FREERDP2_PLUGIN_DIR=${placeholder "out"}/lib"
    patchShebangs ./src/protocols/rdp/plugins/generate-entry-wrappers.pl
    patchShebangs ./src/protocols/rdp/keymaps/generate.pl
```

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -22360,6 +22360,8 @@ with pkgs;
 
   graph-cli = callPackage ../tools/graphics/graph-cli { };
 
+  guacamole = callPackage ../servers/guacamole { };
```
Comment (dasj19) [inline pkgs/top-level/all-packages.nix:22363]: It seems like there are some issues with openssl 3 so I changed this to
`guacamole = callPackage ../servers/guacamole { openssl = pkgs.openssl_1_1; };`
... to get it working

**Review by dasj19 — CHANGES_REQUESTED** [review]
Please check my comments in order to get it working. Thanks for your contributions so far.

**Comment by uningan** [comment]
Only by reading their github mirror, `1.5.0` close to be released.

**Comment by drupol** [comment]
What's the status of this?

**Comment by onny** [comment]
Sorry not working on this anymore :(

**Comment by drupol** [comment]
That's sad, this would be so useful. What was the status of the PR besides the comments that were done already?

**Comment by onny** [comment]
Just tried to fix the issue and package it but without much success. Didn‘t look that deep into it

**Comment by uningan** [comment]
I'll work on this next days.

**Comment by luxus** [comment]
@oneingan still planing to do this?

**Pull request closed without merging**